Lib/test/support/__init__.py

```diff
@@ -1325,6 +1326,18 @@ def flush_std_streams():
         sys.stderr.flush()
 
 
+class WarningsPrinter:
+    def __init__(self, func: Callable[[str], None]) -> None:
+        self.func = func
+        # bpo-39983: Store the original sys.stderr at Python startup to be able to
+        # log warnings even if sys.stderr is captured temporarily by a test.
+        self.orig_stderr = sys.stderr
+
+    def __call__(self, msg: str) -> None:
+        return self.func(msg)
+
+
+@WarningsPrinter
```

Comment:
Mypy doesn't like assigning attributes to functions, unfortunately; it much prefers the idea of a custom callable class

Reply:
(This is the solution you said you preferred in https://github.com/python/cpython/pull/109382#discussion_r1325540441 :-)